ws/api/endpoints.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Mainflux
+// SPDX-Licence-Identifier: Apache-2.0
+
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/go-zoo/bone"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux/internal/apiutil"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/ws"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, err := decodeRequest(r)
+		if err != nil {
+			encodeError(&req, w, err)
+			return
+		}
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+		req.conn = conn
```

Comment:
Please double-check what happens with the WS connection once the header or response is written to the response writer (in the `encodeError`). I'd assume that connections are closed and released, but you need to do some profiling and see.

Reply:
Since, now, encodeError is only called here for decodeRequest(r), only if there is an error.
The function will call return, before a ws-connection is ever created (before this `conn, err := upgrader.Upgrade(w, r, nil)`)

And in the case that subscribe() fails, we no longer call encodeError, just do `req.conn.close()` and return